_posts/2022-08-16-A-Heavy-Lift.md

```diff
@@ -0,0 +1,123 @@
+
+For our transition to **Kestrel + YARP**, we needed to broaden our measurement to include organic health.  Rather than looking for an absolute bar (say >99.99% success), we needed to compare *“before Kestrel + YARP”* and *“after Kestrel + YARP”* organic success and look for anomalies which would point out potential problems.     
+
+## Challenge: Quick Rollback in Production
+
+With a broadened approach to assessing organic health anomalies caused by a diverse set of applications/frameworks on our platform, we required fast mechanisms to undo our **Kestrel + YARP** transition on a per scale-unit basis; in other words, we needed to be able to *“break glass”* quickly when we encountered problems and return to using IIS/HTTP.SYS.
+
+## The Journey: 100% FrontEndRoles using Kestrel/YARP
+
+With all the context and challenges described, here is how the journey looked like in a picture.  
+
+>**TODO:** Add FE Migration image
+
+![FrontEndRole Migration]({{site.baseurl}}/media/2022/08/FE_Migration.jpg)
+
+As you can see this journey took a lot of time; it took over 6 months to complete.
+
+## The Bugs Encountered
+
+We encountered multiple bugs on our journey to **Kestrel + YARP**. Apart from bugs in our business logic, one of the interesting classes of issue we encountered was the treatment of edge-cases in the HTTP specification. The vast diversity of clients that hit our **FrontEndRole** means that we need to be generous in accepting behavior that may not be exactly to the letter and intent of the spec.
+
+A simple example of one of these cases is when a request has leading newline characters (CR and/or LF). Strictly speaking, this isn’t allowed, but it turns out that there are some clients that send requests that start like:
+
+```aspx-csharp
+\rGET / HTTP/1.1\r\n
+...
+```
+
+This is a case that IIS (and some other servers) allow, but because **Kestrel** historically has taken a fairly strict stance, we saw its parser rejecting requests like this with a `BadHttpRequestException`. Working closely with the ASP.NET Core team, we were able to make **Kestrel** a bit more generous in what it accepts (the example above now works in **Kestrel** in .NET 6.0.5 and newer releases).
+
+Some other *interesting* issues uncovered can be found [here](https://github.com/dotnet/aspnetcore/pull/40833), [here](https://github.com/dotnet/aspnetcore/pull/40633) and [here](https://github.com/microsoft/reverse-proxy/pull/1773).
+
+As a result of investigating and addressing this class of issues, not only have we succeeded in a smooth migration, but we’ve also helped make **Kestrel** a more robust server that can deal with more kinds of *“unusual”* traffic.
```

Comment:
@davidfowl Do you think we should say something here that these changes to deviate from the HTTP spec have no negative impact on the vast majority of clients?

I think someone who is militaristic about following standards could read this section and come away with the impression that Microsoft has "ruined" Kestrel because of Azure.

Reply:
Yes, agreed, we can quote this https://en.wikipedia.org/wiki/Robustness_principle. Also we can say that most of the changes were nuances with the RFC (the PRs show that these things are actually supported in other servers). Non standard things need a flag to change the behavior (i.e strict mode vs liberal mode.).

Reply:
I re-worked this a bit to hopefully address your comments, but let me know if you want more changes.